packages/frontend/src/redux/reducers/sign/index.js

```diff
@@ -25,43 +46,73 @@ const sign = handleActions({
                 .reduce((totalAmount, amount) => totalAmount.add(new BN(amount)), new BN(0)).toString(),
             fees: {
                 transactionFees: '', // TODO: Calculate total fees
-                gasLimit: allActions
-                    .filter(a => Object.keys(a)[0] === 'functionCall')
-                    .map(a => a.functionCall.gas)
-                    .reduce((totalGas, gas) => totalGas.add(gas), new BN(0)).toString(),
+                gasLimit: calculateGasLimit(allActions),
                 gasPrice: '' // TODO: Where to get gas price?
             },
             sensitiveActionsCounter: allActions
                 .filter(a => ['deployContract', 'stake', 'deleteAccount'].indexOf(Object.keys(a)[0]) > -1)
                 .length
         };
     },
-    [signAndSendTransactions]: (state, { error, payload, ready }) => {
+    [handleSignTransactions.pending]: (state) => ({
+        ...state,
+        status: SIGN_STATUS.IN_PROGRESS
+    }),
+    [handleSignTransactions.fulfilled]: (state, { payload }) => ({
+        ...state,
+        status: SIGN_STATUS.SUCCESS,
+        transactionHashes: payload
+    }),
+    [handleSignTransactions.rejected]: (state, { error }) => {
```

Comment:
The FIrst part of the code `retryTxDirection`: is checking the situation that we need to handle, whether gas is too much or too low. And it's set only once in the redux state.

The second part `tryRetryTx`: is checking whether we should try to retry or if the gas limits that we set are reached and we have to break the sign flow with hard error.

The third part `if (tryRetryTx)`: is responsible for changing gas but only if the second part said that we should try that.

Reply:
> I don't think it's good that the rejected handler for handleSignTransactions is now responsible for changing gas amounts for actions in our store

My idea here was that changing the gas is part of the error handling for `handleSignTransactions`, so it should be a part of rejected action - basically, immediately when the error appears, we are checking what the problem was, and handle it in the way that it would allow resolving the problem.

Additionally, on the retry screen, we are showing the amount of gas that would be used for retry. So with the above solution, we can simplify the code, by having the cost of the gas already calculated, so we don't have to do that two times - once for showing it to the user, and a second time when we do the request.

Basically, the user has only two possibilities - retry tx with changed gas, or cancel.

So I thought it's reasonable to do that in `rejected` action. But I can change it no problem, it will be just a bit more complicated solution. Let me know what you prefere.

Reply:
Done.